src/SAWScript/Builtins.hs

```diff
@@ -2205,15 +2204,25 @@ ensureMonadicTerm sc t
       False -> monadifyTypedTerm sc t
 ensureMonadicTerm sc t = monadifyTypedTerm sc t
 
--- | A wrapper for either 'Prover.askMRSolver' or 'Prover.assumeMRSolver' from
--- @MRSolver.hs@: if the first argument is @Just str@, prints out @str@
+-- | A wrapper for either 'Prover.askMRSolver' or 'Prover.refinementTerm' from
+-- @MRSolver.hs@: if the second argument is @Just str@, prints out @str@
 -- followed by an abridged version of the refinement being asked, then calls
--- the given function, returning how long it took to execute
-mrSolver :: (SharedContext -> Prover.MREnv -> Maybe Integer -> [(LocalName, Term)] -> Term -> Term -> IO a) ->
-            Maybe SawDoc -> SharedContext -> [(LocalName, Term)] -> TypedTerm -> TypedTerm ->
-            TopLevel (NominalDiffTime, a)
-mrSolver f printStr sc top_args tt1 tt2 =
+-- the given function. On failure, a string of how long the function took to
+-- run is passed to the third argument and the result is used as the message
+-- for 'fail'. On success, if the fourth argument is @Just strf@, a string of
+-- how long the function took to run is passed to @strf@ and the result is
+-- printed, then regardless, the last argument is called on the result.
+mrSolverH :: SharedContext ->
+             Maybe SawDoc -> (String -> String) -> Maybe (String -> String) ->
+             (SharedContext -> Prover.MREnv -> Maybe Integer ->
+              (Set VarIndex -> Sequent -> TopLevel (SolverStats, SolveResult)) ->
+              SV.SAWRefnset -> [(LocalName, Term)] -> Term -> Term ->
+              TopLevel (Either Prover.MRFailure a)) ->
+             SV.SAWRefnset -> [(LocalName, Term)] -> TypedTerm -> TypedTerm ->
+             (a -> TopLevel b) -> TopLevel b
```

Comment:
This type signature is very complicated and hard to follow. Does it make sense to convert some of these to records with named fields? If so, that might be easier to understand. If not, then attaching comments to the various arguments in the type signature would help (see [this function](https://github.com/GaloisInc/cryptol/blob/cd5d006277790910f5314092a79a6f15a25ab596/src/Cryptol/Testing/Random.hs#L81) for an example of what I mean)

Reply:
Yeah this `mrSolverH` is a mess! I believe my thought process was exactly what you said in your second comment, Eddy: I noticed that I was doing the exact same logic before and after the calls to `Prover.askMRSolver` and `Prover.refinementTerm`, and decided to just abstract all of that out as a separate function. But you make a good point, this would all be easier to understand if this was abstracted out int separate functions which do clearly defined things. I will do that.

Reply:
Okay I broke up this function into smaller pieces, let me know what you guys think!

Reply:
Much better! Thanks!